src/structural_natural_bijections.jl

```diff
@@ -0,0 +1,132 @@
+struct Bijections{P, D}
+    info::D
+end
+
+"""
+    to_natural(b::Bijections, t::Tangent)
+
+Map from a structural `Tangent` to a natural tangent. This transformation shouln't change
+the tangent, just its representation.
+"""
+to_natural(b::Bijections, t::Tangent)
+
+"""
+    to_structural(b::Bijections, n)
+
+Map from an natural tangent `n` to a `Tangent`. This transformation shouldn't change the
+tangent, just its representation.
+"""
+to_structural(b::Bijections, n)
+
+# Bidiagonal
+Bijections(p::P) where {P<:Bidiagonal} = Bijections{P, Char}(p.uplo)
+
+function to_natural(b::Bijections{P}, t::Tangent) where {P<:Bidiagonal}
+    return Bidiagonal(t.dv, t.ev, b.info)
+end
+
+function to_structural(b::Bijections{P}, n::Bidiagonal) where {P<:Bidiagonal}
+    return Tangent{P}(dv=n.dv, ev=n.ev)
+end
+
+# Diagonal
+Bijections(::P) where {P<:Diagonal} = Bijections{P, Nothing}(nothing)
+
+to_natural(b::Bijections{P}, t::Tangent) where {P<:Diagonal} = Diagonal(t.diag)
+
+to_structural(b::Bijections{P}, n::Diagonal) where {P<:Diagonal} = Tangent{P}(diag=n.diag)
+
+# UpperTriangular
+Bijections(::P) where {P<:UpperTriangular} = Bijections{P, Nothing}(nothing)
+
+function to_natural(b::Bijections{P}, t::Tangent) where {P<:UpperTriangular}
+    return UpperTriangular(t.data)
+end
+
+function to_structural(b::Bijections{P}, n::UpperTriangular) where {P<:UpperTriangular}
+    return Tangent{P}(data=n.data)
+end
+
+# LowerTriangular
+Bijections(::P) where {P<:LowerTriangular} = Bijections{P, Nothing}(nothing)
+
+function to_natural(b::Bijections{P}, t::Tangent) where {P<:LowerTriangular}
+    return LowerTriangular(t.data)
+end
+
+function to_structural(b::Bijections{P}, n::LowerTriangular) where {P<:LowerTriangular}
+    return Tangent{P}(data=n.data)
+end
+
+# Symmetric
+Bijections(p::P) where {P<:Symmetric} = Bijections{P, Char}(p.uplo)
+
+function to_natural(b::Bijections{P}, t::Tangent) where {P<:Symmetric}
+    return Symmetric(t.data, Symbol(b.info))
```

Comment:
Is this correct, for things like the gradient of `S[2,1]`? Seems like you'll get `[0 1; 1 0]`.

Reply:
Indeed, this is what would happen. Getting a MWE requires disabling Zygote's own `getindex` adjoint, and running the following:
```julia
using ChainRulesCore
using FiniteDifferences
using LinearAlgebra
using Zygote

function ChainRulesCore.rrule(::typeof(getindex), x::Symmetric, ind::Integer)
    project_x = ProjectTo(x)
    y, pb = rrule(getindex, collect(x), ind)
    function getindex_pullback(dy)
        _, dx, _ = pb(dy)
        return NoTangent(), project_x(unthunk(dx)), NoTangent()
    end
    return y, getindex_pullback
end

# Silly identity function.
foo(x) = getindex([0 x; 0 0], 3)

x = randn()
FiniteDifferences.grad(central_fdm(5, 1), foo, x)
Zygote.gradient(foo, x)


foo_sym(x) = getindex(Symmetric([0 x; 0 0]), 3)

FiniteDifferences.grad(central_fdm(5, 1), foo_sym, x)
Zygote.gradient(foo_sym, x)
```
As expected, everything returns 1. I'm running this on master, so none of this PR is involved.

My feeling is that the above implementation of `getindex` is probably how we would do it at the minute.

Separately (for my own sanity), I implemented a version of `getindex` which produces a `Tangent`:
```julia
structural_getindex(x::Symmetric, p::Int, q::Int) = getindex(x, p, q)

# Assume `uplo = :U`.
function ChainRulesCore.rrule(::typeof(structural_getindex), x::Symmetric, p::Int, q::Int)
    function structural_getindex_pullback(dy)
        ddata = zeros(size(x.data))
        if p > q
            ddata[q, p] = dy
        else
            ddata[p, q] = dy
        end
        return NoTangent(), Tangent{Symmetric}(data=ddata), NoTangent(), NoTangent()
    end
    return structural_getindex(x, p, q), structural_getindex_pullback
end

foo_struct(x) = structural_getindex(Symmetric([0 x; 0 0]), 1, 2)

FiniteDifferences.grad(central_fdm(5, 1), foo_struct, x)
Zygote.gradient(foo_struct, x)
```
everything here outputs `1` as expected.

One final example:
```julia
julia> x = Symmetric([0 randn(); 0 0])
2×2 Symmetric{Float64, Matrix{Float64}}:
  0.0       -0.687609
 -0.687609   0.0

julia> Zygote.gradient(x -> getindex(x, 3), x)[1]
2×2 Symmetric{Float64, Matrix{Float64}}:
 0.0  0.5
 0.5  0.0

julia> Zygote.gradient(x -> structural_getindex(x, 1, 2), x)[1]
(data = [0.0 1.0; 0.0 0.0], uplo = nothing)
```

So it looks possible that we can't have both the current implementation of the structural-natural bijections implemented [here in Zygote](https://github.com/FluxML/Zygote.jl/blob/7f2b16987dc4eab236aabd889444cabdcc6f9caf/src/lib/array.jl#L507) and satisfy the primal + tangent test defined in this PR.

Maybe I'm missing something?

If I'm not missing anything, I suppose we have to choose between the two properties.